src/type.rs

```diff
@@ -299,6 +305,91 @@ impl Verify for TypeObj {
     }
 }
 
+/// A wrapper around [`Ptr<TypeObj>`](TypeObj) with the underlying [Type] statically marked.
+#[derive(Debug)]
+pub struct TypePtr<T: Type>(Ptr<TypeObj>, PhantomData<T>);
+
+#[derive(Error, Debug)]
+#[error("TypePtr mismatch: Constructing {expected} but provided {provided}")]
+pub struct TypePtrErr {
+    pub expected: String,
+    pub provided: String,
+}
+
+impl<T: Type> TypePtr<T> {
+    /// Return a [Ref] to the [Type]
+    /// This borrows from a RefCell and the borrow is live
+    /// as long as the returned [Ref] lives.
+    pub fn deref<'a>(&self, ctx: &'a Context) -> Ref<'a, T> {
+        Ref::map(self.0.deref(ctx), |t| {
+            t.downcast_ref::<T>()
+                .expect("Type mistmatch, inconsistent TypePtr")
+        })
+    }
+
+    /// Create a new [TypePtr] from [`Ptr<TypeObj>`](TypeObj)
+    pub fn new(ptr: Ptr<TypeObj>, ctx: &Context) -> Result<TypePtr<T>> {
```

Comment:
As we don't really create anything (we just wrap the pointer add add typeinfo) how about naming this method `cast` or `try_cast`?

How about adding `impl<T> TryFrom<Ptr<TypeObj>> for TypePtr<T> { ... }` ? Then we can use `into` and `try_into` to convert between the 2 types.

Reply:
>How about adding `impl<T> TryFrom<Ptr<TypeObj>> for TypePtr<T> { ... }` ? Then we can use `into` and `try_into` to convert between the 2 types.

But we need a `ctx: &Context`? or am I missing something. I considered adding our own `TryFromWithCtx` (and a `FromWithCtx`) trait, but wasn't sure if it was worth it.

Reply:
>As we don't really create anything (we just wrap the pointer add add typeinfo) how about naming this method `cast` or `try_cast`?

I agree that `new` isn't appropriate. But I don't like `cast` much either. If we add a `TryFromWithCtx`, we wouldn't need this at all right? I'll take a call based on whether it's good to add that trait. So, going back to my previous comment, should we add that trait?

Reply:
I've renamed the conversion functions to `TypePtr::from_ptr` and `TypePtr::to_ptr` now. I'll see if I can introduce a `FromWithContext` and `TryFromWithContext` later on.

Reply:
Oh, you are right. I didn't notice the context. My bad.
I like from_ptr and to_ptr. Although not sure yet often a trait like TryFromWithContext would really be useful for downstream users. +1 on not introducing the trait yet.